### PR TITLE
Query panic!() to useful diagnostic

### PR DESCRIPTION
**Pull request by RossSmyth**
Changes some more ICEs from bare panic!()s

Adds an `expect_job()` helper method as that is a moral equivalent of what was happening at the uses.

re:#118955

### CODE REVIEW

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 1]
```diff
@@ -44,6 +44,16 @@ enum QueryResult {
     Poisoned,
 }
 
+impl QueryResult {
+    /// Unwraps the query job and if poisoned will raise a [`FatalError`]
+    fn unwrap(self) -> QueryJob {
+        match self {
+            Self::Started(job) => job,
+            Self::Poisoned => FatalError.raise(),
```
Comment (compiler-errors) [inline compiler/rustc_query_system/src/query/plumbing.rs:52]: Why did you change this to `FatalError.raise()`? It should be a `panic!("something descriptive here")`, imo.

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 2]
```diff
@@ -190,10 +197,8 @@ where
         let state = self.state;
         let job = {
             let mut shard = state.active.lock_shard_by_value(&self.key);
-            let job = match shard.remove(&self.key).unwrap() {
-                QueryResult::Started(job) => job,
-                QueryResult::Poisoned => panic!(),
-            };
+            let job = shard.remove(&self.key).unwrap().unwrap();
```
Comment (compiler-errors) [inline compiler/rustc_query_system/src/query/plumbing.rs:200]: This `.unwrap().unwrap()` is more awkard than the previous code, imo.

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 3]
```diff
@@ -44,6 +44,16 @@ enum QueryResult {
     Poisoned,
 }
 
+impl QueryResult {
+    /// Unwraps the query job and if poisoned will raise a [`FatalError`]
+    fn unwrap(self) -> QueryJob {
```
Comment (compiler-errors) [inline compiler/rustc_query_system/src/query/plumbing.rs:49]: This perhaps should be called `expect_job` or something?

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 1 above)
Comment (RossSmyth) [reply compiler/rustc_query_system/src/query/plumbing.rs:52]: Because within this module the QueryJob's are unwrapped four times. Two of them previously did `=> panic!()` and two did `=> FatalError.raise()`.

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 2 above)
Comment (RossSmyth) [reply compiler/rustc_query_system/src/query/plumbing.rs:200]: Yeah it is weird

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 3 above)
Comment (RossSmyth) [reply compiler/rustc_query_system/src/query/plumbing.rs:49]: Yeah that would probably be better.

**Comment by RossSmyth** [comment]
In the same vein as #118955 `FatalError.raise()` is very similar to a non-descriptive panic. So if we want these to have nicer messages then `FatalError.raise()` will probably need the same treatment as `panic!()`. The main difference in handling is that there are a couple instances where panics are caught, and if they have some special handling. I mainly converted these over to normalize them when I was able to.

https://github.com/rust-lang/rust/blob/2a7634047a978207aa1c527ed484ae4ada9bba96/compiler/rustc_query_system/src/dep_graph/mod.rs#L63-L65

**Comment by RossSmyth** [comment]
I see, so there are well defined errors that occur that will lead to a query being poisoned, then when checked they will be `QueryResult::Poisoned`, so then it leads to the panic message and stacktrace, which FatalError does not have.

So as I have it right now it is useful for compiler bugs, but not useful for user-facing errors like those testing as it may be more confusing.

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 4]
```diff
@@ -360,7 +369,9 @@ where
                     // so we just return the error.
                     cycle_error(query, qcx, id, span)
                 }
-                QueryResult::Poisoned => FatalError.raise(),
+                QueryResult::Poisoned => {
+                    panic!("job for query '{}' failed to start and was poisoned", query.name())
```
Comment (RossSmyth) [inline compiler/rustc_query_system/src/query/plumbing.rs:373]: So the question is should it be as it is here, or should it be
```suggestion
                    FatalError.raise()
```

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 5]
```diff
@@ -160,21 +173,18 @@ where
...
             let mut lock = state.active.lock_shard_by_value(&key);
-            match lock.remove(&key).unwrap() {
-                QueryResult::Started(job) => job,
-                QueryResult::Poisoned => panic!(),
-            }
+            lock.remove(&key).unwrap().expect_job(self.query_name)
         };
 
+        // Forget ourself so our destructor won't poison the query
+        mem::forget(self);
```
Comment (compiler-errors) [inline compiler/rustc_query_system/src/query/plumbing.rs:186]: Why did you move this down here? If `expect_job` panics, then it will unwind and execute the destructor, which will cause a second panic.

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 6]
```diff
@@ -44,6 +44,18 @@ enum QueryResult {
     Poisoned,
 }
 
+impl QueryResult {
+    /// Unwraps the query job expecting that it has started.
+    fn expect_job(self, query_name: &'static str) -> QueryJob {
+        match self {
+            Self::Started(job) => job,
+            Self::Poisoned => {
+                panic!("job for query '{}' failed to start and was poisoned", query_name)
```
Comment (compiler-errors) [inline compiler/rustc_query_system/src/query/plumbing.rs:53]: I think this doesn't need the query_name honestly.

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_query_system/src/query/plumbing.rs:373]: This panic is fine

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 7]
```diff
@@ -277,13 +285,14 @@ where
...
                     // poisoned due to a panic instead.
                     let lock = query.query_state(qcx).active.get_shard_by_value(&key).lock();
-                    match lock.get(&key) {
-                        // The query we waited on panicked. Continue unwinding here.
-                        Some(QueryResult::Poisoned) => FatalError.raise(),
-                        _ => panic!(
-                            "query result must in the cache or the query must be poisoned after a wait"
-                        ),
+
+                    if let Some(QueryResult::Poisoned) = lock.get(&key) {
```
Comment (compiler-errors) [inline compiler/rustc_query_system/src/query/plumbing.rs:289]: Why did you change this into an `if let`? A match is easier to read.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 5 above)
Comment (RossSmyth) [reply compiler/rustc_query_system/src/query/plumbing.rs:186]: Oh yeah I meant to move that back up. Thanks.

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 4 above)
Comment (RossSmyth) [reply compiler/rustc_query_system/src/query/plumbing.rs:373]: Your below suggestion solves this issue anyways because the test failure was due to double panic.

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 7 above)
Comment (RossSmyth) [reply compiler/rustc_query_system/src/query/plumbing.rs:289]: https://github.com/rust-lang/rust/blob/0919a5145d412c3ff3ca12e315610417fc1e3364/compiler/rustc_query_system/src/query/plumbing.rs#L284-L294

Does not type check. But can just put the panic back.

**Review by davidtwco — APPROVED** [review]

**Comment by fmease** [comment]
10 commits for a +21/-11 diff is quite a lot. I've already rolled it up but I think that this should've been squashed 🙈

**Comment by compiler-errors** [comment]
Wait yes, please squash this 

@bors r-

**Comment by RossSmyth** [comment]
Okie dokie, forgot about this

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 8]
```diff
@@ -277,11 +284,14 @@ where
                     // We didn't find the query result in the query cache. Check if it was
                     // poisoned due to a panic instead.
                     let lock = query.query_state(qcx).active.get_shard_by_value(&key).lock();
+
                     match lock.get(&key) {
-                        // The query we waited on panicked. Continue unwinding here.
-                        Some(QueryResult::Poisoned) => FatalError.raise(),
+                        Some(QueryResult::Poisoned) => {
+                            panic!("query '{}' not cached due to poisoning", query.name())
```
Comment (Zoxc) [inline compiler/rustc_query_system/src/query/plumbing.rs:290]: This branch now introduces a panic, which is not correct.
